### PR TITLE
Abort a task if schedule shuts down

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
* Abort a task if schedule shuts down, install of reschedule the task

**Why?**
there was no task abort state before

**How did you test it?**
UT

**Potential risks**
N/A

**Is hotfix candidate?**
N/A

### CODE REVIEW

**Review by yycptt — APPROVED** [review]
LGTM. But I kinda think scheduler as a library should not call `Cancel()`, which should be a decision made by the user of scheduler. 

Calling Cancel is safe here as we know scheduler shutdown means shard or host is shutting down, and queue processor logic will reload and re-submit the task on shard reload. But scheduler library doesn't know or supposed to know that, and the decision should be made the queue processor or the task implementation.

**Comment by wxing1292** [comment]
> LGTM. But I kinda think scheduler as a library should not call `Cancel()`, which should be a decision made by the user of scheduler.
> 
> Calling Cancel is safe here as we know scheduler shutdown means shard or host is shutting down, and queue processor logic will reload and re-submit the task on shard reload. But scheduler library doesn't know or supposed to know that, and the decision should be made the queue processor or the task implementation.

another approach is to add another state, e.g. (probably bad name) "rejected" indicating that task should not be re-scheduled again.

the point here is to prevent the following loop:

scheduler shutdown & calling reschedule function on task
task reschedule function call scheduler submit function
scheduler submit function accept the task, see internal state being "shutdown", call reschedule function on task, looping

**Review by yux0 — APPROVED** [review]

**Comment by yycptt** [comment]
yeah I get the main idea, so I stamped the PR.

Following are just some only my personal thought on this.

I think the main question here is who/which component can make the decision that a loaded tasks doesn't need to processed any more. The scheduler, the task itself, the queue processor, or the fx part?

To me, the reason the task can be cancelled on shutdown is we know there's only one scheduler (per host/shard doesn't matter). so the fact that scheduler is shutting down means it's safe to cancel/discard the task. Since it's related to how objects are constructed or composed, I don't think scheduler as a standalone component can make the decision.

Queue processor and task itself is more interesting. Since the queue processor constructor only takes in one scheduler, it knows when that only scheduler got shutdown, processing can be stopped. So it can make the cancel decision.
 
Now since queue processor, who has this knowledge, also creates the task. I think task also inherits this knowledge and can cancel itself when found the scheduler got shutdown. If we agree on this, then I would propose simply call task.Nack(ErrSchedulerShutdown) on scheduler shutdown and the impl of Nack() can cancel the task when seeing the shutdown error.

**Comment by wxing1292** [comment]
I would prefer not overloading Nack function. 
Nack should only mark the task as un-process-able, so task can be put to DLQ
are you ok to add a new API, e.g. Abort() indicating shutdown?

**Comment by yycptt** [comment]
> I would prefer not overloading Nack function. Nack should only mark the task as un-process-able, so task can be put to DLQ are you ok to add a new API, e.g. Abort() indicating shutdown?

yeah, that sounds good.

File: common/tasks/interleaved_weighted_round_robin.go [hunk 1]
```diff
@@ -531,7 +531,7 @@ func (s *InterleavedWeightedRoundRobinScheduler[T, K]) hasRemainingTasks() bool
 	return numTasks > 0
 }
 
-func (s *InterleavedWeightedRoundRobinScheduler[T, K]) rescheduleTasks() {
+func (s *InterleavedWeightedRoundRobinScheduler[T, K]) cancelTasks() {
```
Comment (yycptt) [inline common/tasks/interleaved_weighted_round_robin.go:534]: abortTasks()?

**Review by yycptt — APPROVED** [review]

**Pull request merged**